### ISSUE TITLE
Pub/Sub: 1.2.2 RELEASE autoconfigure failure with spring.cloud.gcp.pubsub.enabled=false

### ISSUE THREAD

**Issue by materemias** [spring-cloud/spring-cloud-gcp#2271]
and yet again as with #2037 now the `pubSubTemplate` is broken when `spring.cloud.gcp.pubsub.enabled=false`. This was working with 1.2.1.

error message:
```
***************************
APPLICATION FAILED TO START
***************************

Description:

Parameter 0 of method pubSubHealthIndicator in org.springframework.cloud.gcp.autoconfigure.pubsub.health.PubSubHealthIndicatorAutoConfiguration required a bean of type 'org.springframework.cloud.gcp.pubsub.core.PubSubTemplate' that could not be found.

The following candidates were found but could not be injected:
	- Bean method 'pubSubTemplate' in 'GcpPubSubAutoConfiguration' not loaded because @ConditionalOnProperty (spring.cloud.gcp.pubsub.enabled) found different value in property 'spring.cloud.gcp.pubsub.enabled'


Action:

Consider revisiting the entries above or defining a bean of type 'org.springframework.cloud.gcp.pubsub.core.PubSubTemplate' in your configuration.
```

@elefeint I have tried using `spring.cloud.gcp.pubsub.reactive.enabled=false`, but this time it did not help.

**Comment by dzou** [comment]
Thanks for the report.

Are you running any code that uses the health indicator stuff? The criteria that triggers our health indicator autoconfiguration is shown below:

```
@Configuration
@ConditionalOnClass({HealthIndicator.class, PubSubTemplate.class})
@ConditionalOnEnabledHealthIndicator("pubsub")
@AutoConfigureBefore(HealthContributorAutoConfiguration.class)
@AutoConfigureAfter(GcpPubSubAutoConfiguration.class)
public class PubSubHealthIndicatorAutoConfiguration {

    ... autoconfig code ...

}
```

Looks like the only condition guarding it is the `@ConditionalOnEnabledHealthIndicator("pubsub")` which [is here](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/actuate/autoconfigure/health/ConditionalOnEnabledHealthIndicator.html). I guess if you have the health indicator and pubsub classes on your classpath, this might default to true? Not too familiar with how health indicator works.

@elefeint - Do we need to add a `@ConditionalOnBean(PubSubTemplate.class)` to [PubSubHealthIndicatorAutoconfiguration](https://github.com/spring-cloud/spring-cloud-gcp/blob/master/spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/pubsub/health/PubSubHealthIndicatorAutoConfiguration.java#L48)? I also saw you fixed it with your previous changes to reactive pubsub config; so this might not be the true solution if there is a deeper issue.

**Comment by elefeint** [comment]
@dzou Yes to checking for `ConditionalOnBean` in addition to `ConditionalOnClass`. The former would indicate that base autoconfiguration has been triggered. We could also replicate `ConditionalOnProperty` off switch, but `ConditionalOnBean` seems more reliable, since it would guarantee the needed dependency is present. For the larger issue, it's that we need to get into a habit, when reviewing autoconfiguration changes, to think through expected behavior when required dependencies go missing.

@materemias This time the issue is not related to reactive functionality. Although you are right, it's similar in spirit in that we did not prevent the Pub/Sub health indicator from autoconfiguring when Pub/Sub overall is turned off.

**Comment by materemias** [comment]
@dzou thanks for the prompt reply, We use PubSubTemplate and Actuator which can possibly trigger this behavior, but no direct health indicators afaik.

@elefeint thanks for sharing the details, actually setting `management.health.defaults.enabled=false
` in dev profile solved the issue for the moment, so I at least have a workaround. 

I am sure many use `spring.cloud.gcp.pubsub.enabled=false` in development profiles, please check that condition before releases :)

**Comment by meltsufin** [comment]
I just looked at the original PR, and we had a [comment](https://github.com/spring-cloud/spring-cloud-gcp/pull/2040#discussion_r348034004) related to this that slipped through the cracks.

More importantly, reference documentation was never added for this feature. We can at least describe how to turn it on and off there.

**Comment by elefeint** [comment]
@materemias This fix will go into the 1.2.3 release in May. 

I am glad you found a good workaround; another workaround if you'd like to keep non-Pub/Sub healthcheck endpoints active is to exclude the offending autoconfiguration class explicitly in a property file specific to dev profile:

```
spring.autoconfigure.exclude=org.springframework.cloud.gcp.autoconfigure.pubsub.health.PubSubHealthIndicatorAutoConfiguration
```
